File: src/ingest/cgm_csv.py
"""CGM history from a vendor CSV export: LibreView and Dexcom Clarity.

Both apps let a person download their own archive, and that archive is the
cheapest possible source of readings: thousands of points, already timestamped,
no screenshots to recognise. What the two formats share is exactly one idea —
a row per reading — so the parser is a per-format header map plus one shared
row loop.

The vendors' own quirks the parser has to survive:

* LibreView writes a title line *above* the header, and its glucose column is
  named for the unit the account uses (`Historic Glucose mmol/L` or `… mg/dL`);
  `Record Type` separates the sensor's own 15-minute history (0) from manual
  scans (1) and finger sticks (2).
* Clarity keeps its metadata in data rows (`Event Type` = `FirstName`, `Device`)
  with an empty timestamp, and writes out-of-range readings as the words `Low`
  and `High` rather than numbers.
* Neither is honest about the date order: the same export is `15-01-2024` in
  Europe and `01/15/2024` in the US. Ambiguity is resolved per file, not per
  row — see `_pick_date_order`.

The values themselves are never trusted: anything outside the physiological
range is dropped, the same as a misread screenshot (`spec/ingest.md` § Единицы).
"""

from __future__ import annotations

import csv
import io
import re
from dataclasses import dataclass, field
from datetime import datetime

from src.ingest.units import MGDL, MMOL, is_plausible, to_mmol
from src.logging_setup import get_logger
from src.vision.schemas import GlucoseDraft

log = get_logger("ingest.cgm_csv")

LIBREVIEW = "libreview"
CLARITY = "clarity"

#: A person cannot review 30k rows, and neither can the confirmation step.
MAX_ROWS = 20000

# LibreView `Record Type` values that are a glucose reading.
_LIBRE_HISTORIC = "0"
_LIBRE_SCAN = "1"
_LIBRE_STRIP = "2"
_LIBRE_GLUCOSE_TYPES = {_LIBRE_HISTORIC, _LIBRE_SCAN, _LIBRE_STRIP}

# Day-first and month-first spellings of the same instant, tried in the order
# `_pick_date_order` decides on.
_DAY_FIRST = ("%d-%m-%Y %H:%M", "%d/%m/%Y %H:%M", "%d.%m.%Y %H:%M",
              "%d-%m-%Y %I:%M %p", "%d/%m/%Y %I:%M %p")
_MONTH_FIRST = ("%m-%d-%Y %H:%M", "%m/%d/%Y %H:%M",
                "%m-%d-%Y %I:%M %p", "%m/%d/%Y %I:%M %p")
_UNAMBIGUOUS = ("%Y-%m-%dT%H:%M:%S", "%Y-%m-%d %H:%M:%S", "%Y-%m-%d %H:%M",
                "%Y/%m/%d %H:%M", "%Y/%m/%d %H:%M:%S", "%d-%m-%Y %H:%M:%S")


@dataclass(slots=True)
class CsvImport:
    """What one file turned into. `readings` are ready for `repo.save_glucose`."""

    source: str                       # libreview|clarity
    device: str | None = None
    unit_input: str = MMOL
    readings: list[GlucoseDraft] = field(default_factory=list)
    skipped_rows: int = 0             # rows that are not a glucose reading
    rejected: int = 0                 # readings outside the physiological range
    truncated: bool = False           # the file had more than MAX_ROWS readings

    @property
    def span(self) -> tuple[datetime, datetime] | None:
        if not self.readings:
            return None
        stamps = [r.measured_at for r in self.readings]
        return min(stamps), max(stamps)


class UnknownFormat(Exception):
    """Neither vendor's header was recognised — nothing was parsed."""


def parse_cgm_csv(data: bytes, *, max_rows: int = MAX_ROWS) -> CsvImport:
    """Vendor CSV → readings. Raises `UnknownFormat` when the header is foreign."""
    rows = list(csv.reader(io.StringIO(_decode(data))))
    header_idx, header = _find_header(rows)
    if header is None:
        raise UnknownFormat("не нашёл строку заголовков")
    columns = {_norm(name): idx for idx, name in enumerate(header)}
    body = rows[header_idx + 1 :]
    if _is_libreview(columns):
        return _parse_libreview(columns, body, max_rows=max_rows)
    if _is_clarity(columns):
        return _parse_clarity(columns, body, max_rows=max_rows)
    raise UnknownFormat("не похоже ни на LibreView, ни на Dexcom Clarity")


# ------------------------------------------------------------------ formats

def _is_libreview(columns: dict[str, int]) -> bool:
    return "record type" in columns and any(k.startswith("historic glucose") for k in columns)


def _is_clarity(columns: dict[str, int]) -> bool:
    return "event type" in columns and any(k.startswith("glucose value") for k in columns)


def _parse_libreview(
    columns: dict[str, int], body: list[list[str]], *, max_rows: int
) -> CsvImport:
    historic = _column(columns, "historic glucose")
    scan = _column(columns, "scan glucose")
    strip = _column(columns, "strip glucose")
    unit = _unit_of(historic[0] if historic else "")
    out = CsvImport(source=LIBREVIEW, unit_input=unit)
    stamp_at = columns.get("device timestamp")
    type_at = columns.get("record type")
    device_at = columns.get("device")
    order = _pick_date_order(_cells(body, stamp_at))

    by_type = {
        _LIBRE_HISTORIC: historic[1] if historic else None,
        _LIBRE_SCAN: scan[1] if scan else None,
        _LIBRE_STRIP: strip[1] if strip else None,
    }
    for row in body:
        record_type = _cell(row, type_at)
        if record_type not in _LIBRE_GLUCOSE_TYPES:
            out.skipped_rows += 1
            continue
        value_at = by_type.get(record_type)
        raw = _cell(row, value_at)
        measured_at = _parse_stamp(_cell(row, stamp_at), order)
        if measured_at is None or not raw:
            out.skipped_rows += 1
            continue
        if out.device is None:
            out.device = _cell(row, device_at) or None
        _append(out, measured_at, raw, unit, max_rows=max_rows)
        if out.truncated:
            break
    return out


def _parse_clarity(
    columns: dict[str, int], body: list[list[str]], *, max_rows: int
) -> CsvImport:
    value = _column(columns, "glucose value")
    unit = _unit_of(value[0] if value else "")
    out = CsvImport(source=CLARITY, unit_input=unit)
    stamp_at = next((idx for key, idx in columns.items() if key.startswith("timestamp")), None)
    type_at = columns.get("event type")
    device_at = columns.get("source device id")
    order = _pick_date_order(_cells(body, stamp_at))

    for row in body:
        # Clarity keeps account metadata in data rows; only EGV is a reading
        # the sensor produced, and only it has a timestamp.
        if _cell(row, type_at).upper() != "EGV":
            out.skipped_rows += 1
            continue
        measured_at = _parse_stamp(_cell(row, stamp_at), order)
        raw = _cell(row, value[1]) if value else ""
        if measured_at is None or not raw:
            out.skipped_rows += 1
            continue
        if out.device is None:
            out.device = _cell(row, device_at) or None
        _append(out, measured_at, raw, unit, max_rows=max_rows)
        if out.truncated:
            break
    return out


# ------------------------------------------------------------------ helpers

def _append(
    out: CsvImport, measured_at: datetime, raw: str, unit: str, *, max_rows: int
) -> None:
    if len(out.readings) >= max_rows:
        out.truncated = True
        return
    value = _number(raw, unit)
    if value is None or not is_plausible(value, unit):
        out.rejected += 1
        return
    out.readings.append(
        GlucoseDraft(
            measured_at=measured_at,
            value_mmol=to_mmol(value, unit),
            unit_input=unit,
            device=out.device,
        )
    )


def _number(raw: str, unit: str) -> float | None:
    """A cell to a number. `Low`/`High` are Dexcom's words for the sensor's edges."""
    text = raw.strip().replace(",", ".")
    low = text.lower()
    if low in {"low", "lo"}:
        return 2.2 if unit == MMOL else 40.0
    if low in {"high", "hi"}:
        return 22.2 if unit == MMOL else 400.0
    try:
        return float(text)
    except ValueError:
        return None


def _unit_of(column_name: str) -> str:
    return MGDL if "mg/dl" in column_name.lower() else MMOL


def _column(columns: dict[str, int], prefix: str) -> tuple[str, int] | None:
    for name, idx in columns.items():
        if name.startswith(prefix):
            return name, idx
    return None


def _find_header(rows: list[list[str]]) -> tuple[int, list[str] | None]:
    """LibreView puts a title line above the header; Clarity starts with it."""
    for idx, row in enumerate(rows[:5]):
        norm = {_norm(cell) for cell in row}
        if "record type" in norm or "event type" in norm:
            return idx, row
    return 0, None


def _pick_date_order(samples: list[str]) -> tuple[str, ...]:
    """Day-first or month-first, decided once per file.

    A single row cannot tell `03-04` apart; a file usually can — some row in it
    has a day past the twelfth. Nothing decisive found → day-first, because
    both exports are written in the account's locale and the bot's audience is
    not in the US.
    """
    for text in samples:
        match = re.match(r"\s*(\d{1,2})[-/.](\d{1,2})[-/.](\d{4})", text)
        if not match:
            continue
        first, second = int(match.group(1)), int(match.group(2))
        if first > 12:
            return _DAY_FIRST
        if second > 12:
            return _MONTH_FIRST
    return _DAY_FIRST


def _parse_stamp(text: str, order: tuple[str, ...]) -> datetime | None:
    """Naive local time, as the vendor wrote it — `repo` attaches the zone."""
    value = text.strip()
    if not value:
        return None
    for fmt in (*_UNAMBIGUOUS, *order):
        try:
            return datetime.strptime(value, fmt)
        except ValueError:
            continue
    return None


def _cell(row: list[str], idx: int | None) -> str:
    if idx is None or idx >= len(row):
        return ""
    return row[idx].strip()


def _cells(body: list[list[str]], idx: int | None) -> list[str]:
    return [_cell(row, idx) for row in body if _cell(row, idx)]


def _norm(name: str) -> str:
    return name.strip().lower().lstrip("﻿")


def _decode(data: bytes) -> str:
    """Vendor exports arrive as UTF-8 (with BOM) or Windows-1251 from RU apps."""
    for encoding in ("utf-8-sig", "utf-8", "cp1251"):
        try:
            return data.decode(encoding)
        except UnicodeDecodeError:
            continue
    return data.decode("utf-8", errors="replace")


__all__ = [
    "CLARITY",
    "LIBREVIEW",
    "MAX_ROWS",
    "CsvImport",
    "UnknownFormat",
    "parse_cgm_csv",
]

File: web/app/js/cgmcsv.js
/**
 * cgmcsv.js — импорт истории CGM из выгрузок LibreView и Dexcom Clarity.
 * Зеркало `src/ingest/cgm_csv.py` бота (паритет tg/web, CLAUDE.md #11):
 * те же форматы, тот же выбор порядка дат на весь файл, те же границы
 * правдоподобия. Файл никуда не уходит — разбирается в браузере.
 * spec: spec/web.md § Импорт CGM.
 */

const MG_DL_PER_MMOL = 18.0182;
const MMOL_RANGE = [1.0, 33.3];
const MGDL_RANGE = [18.0, 600.0];

export const MAX_ROWS = 20000;
export const LIBREVIEW = 'libreview';
export const CLARITY = 'clarity';

// LibreView: 0 — история сенсора, 1 — сканирование, 2 — тест-полоска.
const LIBRE_VALUE_COLUMN = { '0': 'historic glucose', '1': 'scan glucose', '2': 'strip glucose' };

export class UnknownFormat extends Error {}

/** Одна строка CSV с учётом кавычек. */
function splitRow(line) {
    const out = [];
    let cell = '';
    let quoted = false;
    for (let i = 0; i < line.length; i++) {
        const ch = line[i];
        if (quoted) {
            if (ch === '"' && line[i + 1] === '"') { cell += '"'; i++; }
            else if (ch === '"') quoted = false;
            else cell += ch;
        } else if (ch === '"') quoted = true;
        else if (ch === ',') { out.push(cell); cell = ''; }
        else cell += ch;
    }
    out.push(cell);
    return out;
}

const norm = (s) => String(s || '').replace(/^﻿/, '').trim().toLowerCase();

/** Заголовок: у LibreView над ним ещё строка названия, у Clarity он первый. */
function findHeader(rows) {
    for (let i = 0; i < Math.min(rows.length, 5); i++) {
        const cells = rows[i].map(norm);
        if (cells.includes('record type') || cells.includes('event type')) return i;
    }
    return -1;
}

function columnStarting(columns, prefix) {
    for (const [name, idx] of Object.entries(columns)) {
        if (name.startsWith(prefix)) return { name, idx };
    }
    return null;
}

/**
 * День вперёд или месяц вперёд — решается один раз на файл: одна строка
 * `03-04` неразличима, а в файле обычно найдётся число больше двенадцатого.
 * Ничего решающего нет — день вперёд: выгрузка пишется в локали аккаунта.
 */
function pickDateOrder(samples) {
    for (const text of samples) {
        const m = /^\s*(\d{1,2})[-/.](\d{1,2})[-/.](\d{4})/.exec(text);
        if (!m) continue;
        if (+m[1] > 12) return 'day';
        if (+m[2] > 12) return 'month';
    }
    return 'day';
}

/** Naive local time → Date в часовом поясе браузера. */
function parseStamp(text, order) {
    const value = String(text || '').trim();
    if (!value) return null;
    const iso = /^(\d{4})-(\d{2})-(\d{2})[T ](\d{2}):(\d{2})(?::(\d{2}))?$/.exec(value);
    if (iso) return new Date(+iso[1], +iso[2] - 1, +iso[3], +iso[4], +iso[5], +(iso[6] || 0));
    const ymd = /^(\d{4})\/(\d{2})\/(\d{2})\s+(\d{1,2}):(\d{2})(?::(\d{2}))?$/.exec(value);
    if (ymd) return new Date(+ymd[1], +ymd[2] - 1, +ymd[3], +ymd[4], +ymd[5], +(ymd[6] || 0));
    const m = /^(\d{1,2})[-/.](\d{1,2})[-/.](\d{4})\s+(\d{1,2}):(\d{2})(?::(\d{2}))?\s*(AM|PM)?$/i.exec(value);
    if (!m) return null;
    const day = order === 'day' ? +m[1] : +m[2];
    const month = order === 'day' ? +m[2] : +m[1];
    let hour = +m[4];
    const suffix = (m[7] || '').toUpperCase();
    if (suffix === 'PM' && hour < 12) hour += 12;
    if (suffix === 'AM' && hour === 12) hour = 0;
    const date = new Date(+m[3], month - 1, day, hour, +m[5], +(m[6] || 0));
    return Number.isNaN(date.getTime()) ? null : date;
}

/** `Low`/`High` — слова Dexcom для краёв диапазона, а не пропуск. */
function toNumber(raw, unit) {
    const text = String(raw || '').trim().replace(',', '.');
    const low = text.toLowerCase();
    if (low === 'low' || low === 'lo') return unit === 'mg/dL' ? 40 : 2.2;
    if (low === 'high' || low === 'hi') return unit === 'mg/dL' ? 400 : 22.2;
    const value = Number(text);
    return text === '' || Number.isNaN(value) ? null : value;
}

function plausible(value, unit) {
    const [low, high] = unit === 'mg/dL' ? MGDL_RANGE : MMOL_RANGE;
    return value >= low && value <= high;
}

function toMmol(value, unit) {
    return Math.round((unit === 'mg/dL' ? value / MG_DL_PER_MMOL : value) * 100) / 100;
}

/**
 * Разбирает файл. Возвращает
 * `{source, unit, device, readings:[{ts, mmol}], skippedRows, rejected, truncated, span}`.
 * Бросает `UnknownFormat`, если заголовок чужой.
 */
export function parseCgmCsv(text, { maxRows = MAX_ROWS } = {}) {
    const rows = text.split(/\r?\n/).filter((line) => line.length).map(splitRow);
    const headerIdx = findHeader(rows);
    if (headerIdx < 0) throw new UnknownFormat('не нашёл строку заголовков');
    const columns = {};
    rows[headerIdx].forEach((name, idx) => { columns[norm(name)] = idx; });
    const body = rows.slice(headerIdx + 1);

    if ('record type' in columns && columnStarting(columns, 'historic glucose')) {
        return parseLibreview(columns, body, maxRows);
    }
    if ('event type' in columns && columnStarting(columns, 'glucose value')) {
        return parseClarity(columns, body, maxRows);
    }
    throw new UnknownFormat('не похоже ни на LibreView, ни на Dexcom Clarity');
}

function emptyResult(source, unit) {
    return { source, unit, device: null, readings: [], skippedRows: 0, rejected: 0, truncated: false, span: null };
}

function cell(row, idx) {
    return idx === undefined || idx === null || idx >= row.length ? '' : String(row[idx]).trim();
}

function push(out, at, raw, unit, maxRows) {
    if (out.readings.length >= maxRows) { out.truncated = true; return; }
    const value = toNumber(raw, unit);
    if (value === null || !plausible(value, unit)) { out.rejected++; return; }
    out.readings.push({ ts: at.toISOString(), mmol: toMmol(value, unit) });
}

function finish(out) {
    if (out.readings.length) {
        const stamps = out.readings.map((r) => r.ts).sort();
        out.span = [stamps[0], stamps[stamps.length - 1]];
    }
    return out;
}

function parseLibreview(columns, body, maxRows) {
    const historic = columnStarting(columns, 'historic glucose');
    const unit = historic && historic.name.includes('mg/dl') ? 'mg/dL' : 'mmol/L';
    const out = emptyResult(LIBREVIEW, unit);
    const stampAt = columns['device timestamp'];
    const typeAt = columns['record type'];
    const deviceAt = columns['device'];
    const order = pickDateOrder(body.map((r) => cell(r, stampAt)).filter(Boolean));

    for (const row of body) {
        const prefix = LIBRE_VALUE_COLUMN[cell(row, typeAt)];
        if (!prefix) { out.skippedRows++; continue; }
        const column = columnStarting(columns, prefix);
        const at = parseStamp(cell(row, stampAt), order);
        const raw = column ? cell(row, column.idx) : '';
        if (!at || !raw) { out.skippedRows++; continue; }
        if (!out.device) out.device = cell(row, deviceAt) || null;
        push(out, at, raw, unit, maxRows);
        if (out.truncated) break;
    }
    return finish(out);
}

function parseClarity(columns, body, maxRows) {
    const value = columnStarting(columns, 'glucose value');
    const unit = value && value.name.includes('mg/dl') ? 'mg/dL' : 'mmol/L';
    const out = emptyResult(CLARITY, unit);
    const stampKey = Object.keys(columns).find((k) => k.startsWith('timestamp'));
    const stampAt = stampKey === undefined ? undefined : columns[stampKey];
    const typeAt = columns['event type'];
    const deviceAt = columns['source device id'];
    const order = pickDateOrder(body.map((r) => cell(r, stampAt)).filter(Boolean));

    for (const row of body) {
        // Clarity держит метаданные аккаунта в строках данных; замер — только EGV.
        if (cell(row, typeAt).toUpperCase() !== 'EGV') { out.skippedRows++; continue; }
        const at = parseStamp(cell(row, stampAt), order);
        const raw = value ? cell(row, value.idx) : '';
        if (!at || !raw) { out.skippedRows++; continue; }
        if (!out.device) out.device = cell(row, deviceAt) || null;
        push(out, at, raw, unit, maxRows);
        if (out.truncated) break;
    }
    return finish(out);
}

export const SOURCE_NAMES = { [LIBREVIEW]: 'LibreView', [CLARITY]: 'Dexcom Clarity' };
